plugins/inputs/redis/redis.go

```diff
@@ -344,6 +391,24 @@ func (r *Redis) gatherCommandValues(client Client, acc telegraf.Accumulator) err
 	return nil
 }
 
+func discoverNodes(client Client) ([]redisClusterNode, error) {
+	val, err := client.Do("string", "cluster", "nodes")
+
+	if err != nil {
+		if strings.Contains(err.Error(), "unexpected type=") {
+			return nil, fmt.Errorf("could not get command result: %w", err)
+		}
+
+		return nil, err
+	}
+
+	str, ok := val.(string)
+	if ok {
+		return parseClusterNodes(str)
+	}
+	return nil, fmt.Errorf("could not discover nodes: %w", err)
```

Comment:
I would prefer
```suggestion
	str, ok := val.(string)
	if !ok {
		return nil, fmt.Errorf("could not discover nodes: %w", err)
	}
	return parseClusterNodes(str)
```
and maybe fold the `parseClusterNodes` function into this one...

Reply:
I liked separating parseClusterNodes to make it more easily/directly testable.  Can you say more about why you prefer to fold it in?

Reply:
(To add to this...I expect maintenance of this function may take some work as/if Redis `cluster nodes` format changes in the future.  As you can see from the tests, there are already 3 variations.)

Reply:
It's mostly a personal preference but splitting code into two functions, i.e. there is only one caller, where the caller side is very small usually requires the reader to jump between different places and makes following the flow more difficult. Like in this example, the `discoverNodes` function reads, "do a API call and parse the result". This adds no information to the reader as you cannot see _how_ and _what_ is parsed.

The splitting can be reasonable IMO if a) there are many processing steps, b) a lot of branches or c) complex processing steps. None of those criteria are applicable here as the parsing is fairly "simple" in terms of code...